### ISSUE TITLE
Document better workflow for working off topic branches

### ISSUE THREAD

**Issue by dcolligan** [ga4gh/server#849]
e.g. long running topic branches like `rna`

**Comment by jeromekelleher** [comment]
Yes please! We are going to have a bunch of these running in parallel for the next few months, so it would be great to have a proper system for keeping them up to date
.

**Comment by jeromekelleher** [comment]
@dcolligan --- would it be possible to make the main topic branches like variation_annotation, rna, etc, protected by GitHub? Is force-pushing necessary in a strategy for keeping the branches up-to-date with master?

**Comment by macieksmuga** [comment]
Here's my proposed writeup, based on @saupchurch and my experiences with the RNA topic branch.
Terminology wise, there will be three classes of branches:

* the TRUNK (which we now call develop, but is usually called master),
* one or more TOPIC BRANCH(es) which is, for all intents and purposes, the trunk for the topic team, and
* FEATURE BRANCHES on which all new feature development occurs.

The TOPIC BRANCH is kept in sync with the TRUNK by the TOPIC BRANCH maintainer, who merges recent updates to TRUNK back into the TOPIC BRANCH.

Any individual developer has a copy of the TOPIC BRANCH in their fork, and keeps it current against the ga4gh:TOPIC_BRANCH

Actual feature development is done on discrete FEATURE BRANCHES, which are branched off the developer’s copy of the TOPIC BRANCH.

The FEATURE BRANCHES are on each developer’s individual forked repo only, they never make it to the ga4gh repo.

When a developer is ready to PR their feature code, they:

a) make sure their fork’s copy of the TOPIC BRANCH is up to date against the GA4GH repo’s version
(git checkout <topic_branch>; git pull upstream <topic_branch>)
b) rebase their FEATURE BRANCH against their updated copy of the TOPIC BRANCH
(git checkout <feature_branch>; git rebase <topic_branch>)
c) issue a PR of this rebased FEATURE BRANCH against the ga4gh TOPIC BRANCH

Specific things for feature developers to note are:

- individual developers are not responsible for keeping the TOPIC BRANCH up to date. That is the branch maintainer’s job
- rebasing by developers is done ONLY from their personal FEATURE BRANCHES onto their copy of the TOPIC BRANCH. (once they’re ready to issue a PR, that is),

**Comment by macieksmuga** [comment]
@jeromekelleher force pushing shouldn't be necessary. Keeping topic branches up to date with master should involve no more than a simple fast-forward merge on the branch maintainer's side, followed by a regular push upstream:
```
git checkout <topic branch>; git pull upstream <topic branch>
git pull upstream master
[ possible merge cleanup here, for example resolving conflicts in _protocol_definitions.py ]
git push upstream <topic branch>
```

**Comment by dcolligan** [comment]
@jeromekelleher that's almost the same question as asking "are we going to be rebasing (to use @macieksmuga 's terminology) topic branches on top of trunk?"  If yes, we can't protect them, since rebasing rewrites history in the branch and thus must use force pushing to update the branch.  If no, we have the option of protecting them.

@macieksmuga 's writeup seems fine as far as it goes, but it doesn't address two central questions: 1) who are going to be the topic branch maintainers and 2) how are they going to keep the topic branch current with trunk?

**Comment by macieksmuga** [comment]
Good questions, @dcolligan, here are my answers to them:

1) to be decided per-branch. For the time being, I'm on point for RNA/sequence annotations and David Steinberg is handing Variation annotation. 

2) The steps for a branch maintainer to keep topic branch current with trunk are as follows:

* on latest local copy of topic branch, merge in latest version of master (`git pull upstream master` usually does the trick, but I'm sure there are other ways that provide more control over the process)
* resolve any merge conflicts - much of the time these will be minimal, but the occasional large change to master (hi protocol buffers!) will of course necessitate some real work here.
* note in the merge's commit message what commit from master the branch was updated to
* push updated local topic branch up to its upstream (ga4gh).

**Comment by macieksmuga** [comment]
Also note that in the scheme I proposed, I'd stay away from rebasing except in the one case specified above - ie. in developers' individual feature branches. Topic branches would be kept in sync with master through merging only.

**Comment by dcolligan** [comment]
So here are my thoughts on question (2):

There are three possible git workflows that I can conceive of for scenarios in which (like ours) there is a multilevel hierarchy of branches in which downstream branches submit changes to upstream branches, and in which downstream branches need to continuously incorporate updates from upstream branches.

Workflow A: "Pure rebase": every downstream branch submits to its upstream branch by rebasing on top of the upstream branch and submitting a pull request.  Upstream branch maintainers update an upstream branch by rebasing against the upstream branch relative to it, then force pushing the upstream branch and notifying everyone downstream of the update.  Developers update an upstream branch locally by deleting the branch locally and then pulling it from the repo.  

- the only branch that can be protected is the trunk branch
- lots of force pushing done both remotely and locally
- linear commit history, both for trunk and upstream branches

Workflow B: "Pure merge": every downstream branch submits to its upstream branch by pull request, without rebasing.  Upstream branch maintainers update an upstream branch by merging from the upstream branch relative to it (and perhaps notifying everyone downstream, although this is more implicit in the git history).  Developers update an upstream branch locally by fast-forward merging from the remote version of the branch.  This is basically http://nvie.com/posts/a-successful-git-branching-model/

- can protect both trunk and upstream branches
- non-linear commit history

Workflow C: "Merge, except rebase at the leaves [or some other well-defined point in each branch hierarchy]": this is the same as B, except that the branches that don't have anyone downstream from them (i.e. individual developers' feature branches) submit by rebasing on top of their immediate upstream branch and creating a PR.

- can protect both trunk and upstream branches
- non-linear commit history (a bit less so than B)

So basically our decision comes down to whether we think that merging or rebasing is more awesome for the project.  In some ways we have already made this decision, coming down on the side of rebasing.  The additional wrinkle with multilevel branch hierarchies (as opposed to master/feature branch hierarchies of H = 2) is that if we go with A developers need to know how to update their upstream branches by blowing away/recreating instead of fetch/fast forward merge.  Also, upstream branch maintainers need to be confident rebasing and force pushing their topic branches.  Nevertheless, if we were going to make a project-wide decision, I would advocate Workflow A.

I suppose each topic branch team could decide what workflow works for them (perhaps stipulating that submissions to master should be done in a certain way?).  But there could be value in standardizing our workflows.  I'm ambivalent on this flexibility vs standardization question.

**Comment by macieksmuga** [comment]
I'm strongly advocating for C. A very good discussion on this issue can be found here:
https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/
which incidentally omits B as an option, since rebasing for local cleanup is a desirable and acceptable thing, no matter whether the shared team policy is rebase or merge).

A major reason I'm against A is that it doesn't play nicely with pull requests. Quoting from the above-linked discussion:

> Rebasing doesn't play well with pull requests, because you can't see what minor changes 
> someone made if they rebased (incidentally, the consensus inside the Bitbucket Server 
> development team is to never rebase during a pull request).

This seems to run counter to the development process we've developed so far, if I understand it correctly.

**Comment by diekhans** [comment]
I guess I really don't care so much about code history so much as what gets us working code in the easiest and safest manner.   It the revision history is a bit of a mess, is that an important thing to spend time on avoiding?

**Comment by diekhans** [comment]
@dcolligan you are our git guru. Ignoring previous directions, what do you recommend?  It does seem like C is the best trade-off, but then again, I am a gitiot

**Comment by dcolligan** [comment]
@macieksmuga that's a good article, but doesn't address the multilevel issue.  Granted, it could be argued that mutilevel strategies are just logical extensions of 2-level strategies.

The author argues that the big pro of merge-based strategy is traceability of branches.  I've never understood why some think it's important to know that e.g. a feature was developed in branch `awesome-feature-1234`.  It seems like superfluous information to me.  The downside of merge-based strategies is that you can get a log graph that looks like that picture (shudder) and there are a lot of git tools that make much more sense with a more linear history: bisect, log, etc.  It's also easier to understand temporally what happened to the code, which is important in and of itself, and also important for doing things like reverting a change.

We've already been rebasing with pull requests, and it seems to be going fine.  If one wants to avoid the "not seeing minor changes" issue, one can commit fix-it commits to one's PR branch incrementally during review, and squash the commits at the end.

**Comment by dcolligan** [comment]
@diekhans Our decision has to be driven by the following considerations:

1) Who is going to be managing the topic branches?  How comfortable with git are they?
2) Do we want to enforce a workflow for topic branches, or let topic branch managers decide how they want to manage their topic branch and contributions to it from downstream?

If we have people who feel comfortable with git and rebasing, and we want to enforce workflow globally, we should do A.  If we don't have people who feel comfortable with git, and we want to enforce workflow globally, we should do C.  If we don't care about enforcing workflow globally, each topic branch manager can decide how he or she is going to manage the topic branch and contributors.

**Comment by diekhans** [comment]
thanks for your wisdom @dcolligan.   If you get beyond the basics, git is really steep learning curve.  We don't necessarily have the expertise in the task teams.   I also don't see the importance of global workflow across the topic branches.  We more need to do this in a a way that works best for the people doing the work.  Git pain has wasted a lot of time in a couple of instances.

We just need to give guidance.  I would say we suggest C (and give a simple cookbook) unless a group wants a more powerful approach.

**Comment by jeromekelleher** [comment]
This is difficult. Thanks @dcolligan and @macieksmuga for laying out the alternatives clearly. Here are my thoughts.

1. I like the current rebase-for-PRs to master approach. It make the history nice and easy to follow, and this is important.
2. I don't think we need a fully linear commit history when the topic branch is ultimately merged into master. Since we are doing things in parallel, it seems inevitable that there will be some non-linearity.

So, I guess I'm voting for C also. The remaining thing is, should the merges be done directly on the command line, or with, say, weekly merge PRs? Most of the time these will be simple, but sometimes we'll have messes to clean up and it would be good to have visibility.

**Comment by diekhans** [comment]
I would much rather have merges being done directly based on
judgment of difficult and priority.  Weekly is far to infrequent
for simple merges and one might have to clear a few hours for a
complex one.

**Comment by jeromekelleher** [comment]
Weekly doesn't seem that long --- we'd hope that the topic branches and the updates to master are happening on different parts of the codebase, and don't collide. I think we'd be asking a lot of the topic branch maintainer if they have to pull in every single commit that happens on master into the topic branch as it happens.

**Comment by diekhans** [comment]
Oh, I wasn't implying that things happen immediately either.  It
happens as needed.  If a feature fork needs something ASAP from
master and we are requiring them to go via the topic branch to
get it, we probably want to rebase ASAP.


Weekly could be a fine guideline, just not a hard and fast
requirement.

**Comment by dcolligan** [comment]
Fixed in #862
